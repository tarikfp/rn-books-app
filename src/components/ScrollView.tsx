/* eslint-disable @typescript-eslint/no-unused-vars */
import * as React from 'react';
import { ScrollView } from 'react-native';
import { s } from '~theme/common-styles';
import { makeStyles } from '~theme/make-styles';

/**
 * Reusable custom scroll view component that can be used throughout the app
 * For the sake of this example, it will be used only in SearchResultDetailScreen
 */
interface Props extends React.ComponentProps<typeof ScrollView> {
  backgroundColor?: string;
}

const AppScrollViewInner: React.ForwardRefRenderFunction<
  any,
  React.PropsWithChildren<Props>
> = ({ backgroundColor, testID, children, ...props }, scrollViewRef) => {
  const styles = useStyles();

  return (
    <ScrollView
      ref={scrollViewRef}
      keyboardShouldPersistTaps="handled"
      style={styles.base}
      scrollEventThrottle={16}
      contentContainerStyle={styles.contentContainer}
      testID={testID}
      {...props}
    >
      {children}
    </ScrollView>
  );
};

export const AppScrollView = React.forwardRef(AppScrollViewInner);

const useStyles = makeStyles((theme) => ({
  contentContainer: {
    ...s.flexGrow,
    backgroundColor: theme.colors.background,
  },
  base: {
    ...s.flex,
    backgroundColor: theme.colors.background,
  },
}));
