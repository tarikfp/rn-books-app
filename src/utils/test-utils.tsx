import { ThemeProvider } from '@shopify/restyle';
import { render, RenderOptions } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { theme } from '~theme/core';
import { ThemePreferenceContext } from '~theme/useThemePreference';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemePreferenceContext.Provider
      value={{
        isDarkTheme: false,
        toggleTheme: () => {},
      }}
    >
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemePreferenceContext.Provider>
  );
};

const withAllTheProviders = (ui: React.ReactNode) => (
  <AllTheProviders children={ui} />
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderer = (ui: React.ReactElement) =>
  renderer.create(withAllTheProviders(ui));

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
// override renderer method
export { customRenderer as createRenderer };
