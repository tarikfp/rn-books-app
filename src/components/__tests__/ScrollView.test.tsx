import React from 'react';
import { Text } from 'react-native';
import { AppScrollView } from '~components/ScrollView';
import { render } from '~utils/test-utils';

describe('AppScrollView', () => {
  it('renders correctly with default props', () => {
    const { toJSON } = render(
      <AppScrollView>
        <Text>Test Content</Text>
      </AppScrollView>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
