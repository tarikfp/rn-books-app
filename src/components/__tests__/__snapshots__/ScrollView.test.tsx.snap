// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppScrollView renders correctly with default props 1`] = `
<RCTScrollView
  contentContainerStyle={
    {
      "backgroundColor": "#F3F4F6",
    }
  }
  keyboardShouldPersistTaps="handled"
  scrollEventThrottle={16}
  style={
    {
      "backgroundColor": "#F3F4F6",
    }
  }
>
  <View>
    <Text>
      Test Content
    </Text>
  </View>
</RCTScrollView>
`;
